### ISSUE TITLE
Unsetting http-proxy and https-proxy variables on ec2 provisioning

### ISSUE THREAD

**Issue by eric-w-davidson** [test-kitchen/test-kitchen#1020]
## Description

We are running kitchen on a internal network that requires a proxy for outbound internet communication.  The http_proxy and https_proxy environment variables need set in order to make EC2 outbound provisioning (via kitchen-ec2 driver) calls to create a new EC2 instance.  After the EC2 instance spins up, we are running our chef recipes and making URL requests, however once the instance is in Amazon we do not want the proxy set.  We only need it set to provision the EC2 instance and not set when the Chef recipes are running.  Is there a way of instructing kitchen to not set/pass our environment variables to the newly created instances?  

Below is the details/error we are receiving.

## Kitchen Version

Tell us which version of test-kitchen you are using (`kitchen --version`).  

```ruby
Test Kitchen version 1.7.2
```

## ChefDK Version

If you are running test-kitchen via ChefDK, `chef --version` will provide additional relevent version details.

```ruby
$ chef --version
Chef Development Kit Version: 0.12.0
chef-client version: 12.8.1
berks version: 4.3.2
kitchen version: 1.7.2```

## Ruby Version
ruby 2.1.6p336 (2015-04-13 revision 50298) [x86_64-darwin13.0]

## Platform Version

Tell us which Operating System distribution and version test-kitchen is running on.

Mac OSX 10.10.5

Darwin 14.5.0 Darwin Kernel Version 14.5.0: Tue Sep  1 21:23:09 PDT 2015; root:xnu-2782.50.1~1/RELEASE_X86_64 x86_64
## Replication Case

Tell us what steps to take to replicate your problem.  See [How to create a Minimal, Complete, and Verifiable example](https://stackoverflow.com/help/mcve)
for information on how to create a good replication case.

1.)  Set environment variables pointing to a internal proxy (http_proxy and https_proxy) that is not reachable via AWS and run using kitchen-ec2 driver.
2.)  Have a chef cookbook that connect to a internet service
3.)  Execute kitchen converge
4.)  Error should pop up stating it can't reach proxy.

## Kitchen Output

The relevant output of the test-kitchen run or a link to a gist of the entire run, if there is one.

The debug output (kitchen -l debug) may be useful, but please link to a gist, or truncate it.

Please do not paste the full run out put here.  If the full output would be useful, link to it in a gist.

```ruby
>>>>>> Converge failed on instance <default-amazon>.
>>>>>> Please see .kitchen/logs/default-amazon.log for more details
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: SSH exited (1) for command: [**sh -c '
http_proxy="xxx-proxy.domain:80"; export http_proxy
https_proxy="http://xxx-proxy.domain:80"; export https_proxy**
sudo -E /opt/chef/bin/chef-client --local-mode --config /tmp/kitchen/client.rb --log_level auto --force-formatter --no-color --json-attributes /tmp/kitchen/dna.json --chef-zero-port 8889
```

## Kitchen Diagnose

Please include the output of `kitchen diagnose --all` or `kitchen diagnose --loader` (if the first failed).  This can be helpful in troubleshooting, so please include it in a gist.

$ kitchen diagnose --all
---
timestamp: 2016-05-03 14:13:45 UTC
kitchen_version: 1.7.2
plugins:
  driver:
    Ec2:
      class: Kitchen::Driver::Ec2
      version: 1.0.0
      api_version: 2
  provisioner:
    ChefZero:
      class: Kitchen::Provisioner::ChefZero
      version: 1.7.2
      api_version: 2
  transport:
    Ssh:
      class: Kitchen::Transport::Ssh
      version: 1.7.2
      api_version: 1
  verifier:
    Busser:
      class: Kitchen::Verifier::Busser
      version: 1.7.2
      api_version: 1
loader:
  process_erb: true
  process_local: true
  process_global: true
  global_config: 
  project_config:
    filename: "<path>/av-host-auth/.kitchen.yml"
    raw_data:
      driver:
        name: ec2
        instance_type: t2.small
        tags:
          created-by: '<uid>'
          Name: ut-2fa-local
          UAI: UAILAB
      provisioner:
        name: chef_zero
      platforms:
      - name: amazon
        driver:
          image_id: <aws-image-id>
      suites:
      - name: default
        run_list:
        - recipe[av-host-auth]
        attributes: 
  local_config:
    filename: "<path>/av-host-auth/.kitchen.local.yml"
    raw_data:
      driver:
        name: ec2
        shared_credentials_profile: temp
        aws_ssh_key_id: <keypair>
        security_group_ids:
        - <aws-sg-id>
        region: us-east-1
        availability_zone: us-east-1b
        require_chef_omnibus: true
        subnet_id: <aws-subnet-id>
      transport:
        ssh_key: "<keypair>"
        ssh_timeout: 10
        ssh_retries: 5
        username: <uid>
      provisioner:
        name: chef_zero
        chef_omnibus_url: http://xxx/chef/nodes/install.sh
        ssl_verify_mode: verify_none
        encrypted_data_bag_secret_key_path: "<path-to-data-bag-secret>"
        data_bags_path: "/Users/data_bags/"
        environments_path: "<path>/.chef/environments/"
  combined_config:
    filename: 
    raw_data:
      driver:
        name: ec2
        instance_type: t2.small
        tags:
          created-by: '<uid>'
          Name: ut-2fa-local
          UAI: <value>
        shared_credentials_profile: temp
        aws_ssh_key_id: <keypair-name>
        security_group_ids:
        - <aws-sg-id>
        region: us-east-1
        availability_zone: us-east-1b
        require_chef_omnibus: true
        subnet_id: <aws-subnet-id>
      provisioner:
        name: chef_zero
        chef_omnibus_url: http://xxx/chef/nodes/install.sh
        ssl_verify_mode: verify_none
        encrypted_data_bag_secret_key_path: "<path-to-data-bag-secret>"
        data_bags_path: "/Users/data_bags/"
        environments_path: "<path>/.chef/environments/"
      platforms:
      - name: amazon
        driver:
          image_id: <aws-ami-id>
      suites:
      - name: default
        run_list:
        - recipe[av-host-auth]
        attributes: 
      transport:
        ssh_key: "<path-to-keypair>"
        ssh_timeout: 10
        ssh_retries: 5
        username: <uid>
instances:
  default-amazon:
    platform:
      os_type: unix
      shell_type: bourne
    state_file:
      hostname: <ip>
      last_action: create
      server_id: <aws-instance-id>
    driver:
      associate_public_ip: 
      availability_zone: us-east-1b
      aws_access_key_id: 
      aws_secret_access_key: 
      aws_session_token: 
      aws_ssh_key_id: <ssh-keypair>
      block_duration_minutes: 
      ebs_optimized: false
      http_proxy: 
      iam_profile_name: 
      image_id: <aws-ami-id>
      image_search: 
      instance_type: t2.small
      interface: 
      kitchen_root: "<path>/av-host-auth"
      log_level: :info
      name: ec2
      private_ip_address: 
      region: us-east-1
      retry_limit: 3
      retryable_sleep: 5
      retryable_tries: 60
      security_group_ids:
      - <aws-sg-id>
      shared_credentials_profile: temp
      spot_price: 
      subnet_id: <aws-subnet-id>
      tags:
        created-by: '<uid>'
        Name: ut-2fa-local
        UAI: UAILAB
      test_base_path: "<path>/av-host-auth/test/integration"
      user_data: 
      username: 
    provisioner:
      attributes: {}
      chef_client_path: "/bin/chef-client"
      chef_omnibus_install_options: 
      chef_omnibus_url: http://xxx/chef/nodes/install.sh
      chef_zero_host: 
      chef_zero_port: 8889
      client_rb: {}
      clients_path: 
      command_prefix: 
      config_path: 
      cookbook_files_glob: README.*,metadata.{json,rb},attributes/**/*,definitions/**/*,files/**/*,libraries/**/*,providers/**/*,recipes/**/*,resources/**/*,templates/**/*
      data_bags_path: "/Users/data_bags"
      data_path: 
      encrypted_data_bag_secret_key_path: "<path-to-data-bag-secret>"
      environments_path: "<path>/.chef/environments"
      ftp_proxy: 
      http_proxy: 
      https_proxy: 
      json_attributes: true
      kitchen_root: "<path>/av-host-auth"
      log_file: 
      log_level: auto
      name: chef_zero
      nodes_path: 
      profile_ruby: false
      require_chef_omnibus: true
      roles_path: 
      root_path: "/tmp/kitchen"
      ruby_bindir: "/embedded/bin"
      run_list:
      - recipe[av-host-auth]
      ssl_verify_mode: verify_none
      sudo: true
      sudo_command: sudo -E
      test_base_path: "<path>/av-host-auth/test/integration"
    transport:
      compression: true
      compression_level: 6
      connection_retries: 5
      connection_retry_sleep: 1
      connection_timeout: 15
      keepalive: true
      keepalive_interval: 60
      kitchen_root: "<path>/av-host-auth"
      log_level: :info
      max_wait_until_ready: 600
      name: ssh
      port: 22
      ssh_key: "<dir-to-keypair>"
      ssh_retries: 5
      ssh_timeout: 10
      test_base_path: "<path>/av-host-auth/test/integration"
      username: <uid>
    verifier:
      busser_bin: "/tmp/verifier/bin/busser"
      chef_omnibus_root: "/opt/chef"
      command_prefix: 
      ftp_proxy: 
      http_proxy: 
      https_proxy: 
      kitchen_root: "<path>/av-host-auth"
      log_level: :info
      name: busser
      root_path: "/tmp/verifier"
      ruby_bindir: "/opt/chef/embedded/bin"
      sudo: true
      sudo_command: sudo -E
      suite_name: default
      test_base_path: "<path>/test/integration"
      version: busser

## NOTE:

This repository is for core issues with the Test-Kitchen framework, so if the issue is with a plugin you may be referred to file the issue with the appropriate plugin.

**Comment by cheeseplus** [comment]
You can do this by setting empty proxy settings as per https://github.com/test-kitchen/test-kitchen/pull/936

**Comment by cheeseplus** [comment]
@eric-w-davidson were you able to get this working with the suggestion above? I'd like to close this issue  but want to make sure you've gotten it working first.

**Comment by moofrank** [comment]
I don't think #936 is working. I'm on Kitchen 1.7.3 (which definitely includes that PR#) , and empty settings are still propagating the local http_proxy and https_proxy environment variables. I'm having to unset my local variables to get converges to work. I'm verifying by doing a kitchen converge -l debug.

**Comment by rhgrant10** [comment]
It seems to me that the default behavior should be to *not* magically pass any settings through to the VM under test. Doing so creates unnecessary differences between the way kitchen behaves and the way chef behaves, which ultimately undermines the paradigm of testing with kitchen. Setting environment variables should be done using chef resources! I could, however, understand having a helpful means of passing a whitelist of environment variables through to the VM, but that should definitely be explicit and "opt-in."

**Comment by cheeseplus** [comment]
@rhgrant10 test-kitchen isn't just for Chef and the http_proxy, https_proxy, no_proxy mechanisms are ubiquitous hence why we honor them by default.

If the behaviour as per #936 is broken that is a regression and we should fix that but I'm also open to making dealing with proxies friendlier in any way we can without breaking existing defaults and behaviour.

**Comment by rhgrant10** [comment]
@cheeseplus that's true; test-kitchen isn't *just* for use with chef. But to be honest, that fact is relatively easy to forget given that chef is the only provisioner mentioned on [Kitchen's home page](http://kitchen.ci/) and the [Kitchen's Getting Started Guide](http://kitchen.ci/docs/getting-started/) has several sections devoted to chef-specific topics.

Regardless, I believe the same paradigm breakdown applies to other provisioners/drivers as well unless they also pass environment variables from one system to another. Please don't hesitate to correct me here, as my use case *is* with chef and my knowledge/use of test-kitchen certainly doesn't extend far past that :)

There is a difference between the fact that proxy settings are found everywhere and the idea that using those proxy settings everywhere is the right thing to do by default. For my use case, I'm developing cookbooks for systems that will not be allowed to reach the internet in an environment that requires proxy settings in order to properly operate. That is, I have to be able to reach the proxy but the systems under test must not be allowed to reach the proxy.

I think the sensible thing to do here is simply to make the behavior of passing environment variables opt-in. Then existing defaults and behavior can be maintained while providing a helpful way of dealing with proxies a breeze. Perhaps using a whitelist of environment variables to pass through could be a simple solution that makes everyone happy.

**Comment by cheeseplus** [comment]
> But to be honest, that fact is relatively easy to forget given that chef is the only provisioner mentioned on Kitchen's home page and the Kitchen's Getting Started Guide has several sections devoted to chef-specific topics.

You're right but we're hoping to address that more directly by pulling Chef out of core https://github.com/test-kitchen/test-kitchen/issues/754 and making shell the default provisioner https://github.com/test-kitchen/test-kitchen/issues/913.

Back to the topic at hand:
From a field perspective, most customers want all their tools to work with proxies automatically using the fairly standard proxy settings - in my experience this is slightly unique from the gamut of potential environmental variables. In the Chef ecosystem that means Berks, TK, Chef, etc all respect those proxy specific system settings as other tools would by default.

I want to balance what I know folks in the field/on the ground expect with what reasonable defaults. We already have some issues/request related to passing through arbitrary env variables so I want to keep this thread on proxies as there is existing code and tests just for this facility.

We can definitely talk about passing through env variables or whitelists but I want to first verify this is a regression and if so fix that one before we add new functionality.

**Comment by rhgrant10** [comment]
> From a field perspective, most customers want all their tools to work with proxies automatically using the fairly standard proxy settings - in my experience this is slightly unique from the gamut of potential environmental variables. In the Chef ecosystem that means Berks, TK, Chef, etc all respect those proxy specific system settings as other tools would by default.

Agreed, proxies are a genuine PITA. However, I think there is a real difference between expecting the tools that create a VM to obey local environment variables and expecting the VM itself to obey the environment variables of the machine that created it without explicit settings. The former is typical and expected, whereas the latter is really only desirable under certain circumstances. The commonality of those circumstances should not imply an expectation since we're talking about a paradigm of creating code to define a separate system that should behave identically regardless of which machine creates it.

> We already have some issues/request related to passing through arbitrary env variables so I want to keep this thread on proxies as there is existing code and tests just for this facility.
> 
> We can definitely talk about passing through env variables or whitelists but I want to first verify this is a regression and if so fix that one before we add new functionality.

Good point, and fair enough :)

**Comment by moofrank** [comment]
Looking at the PR, I can't see why it wouldn't work or why the spec test wouldn't catch it. (My Ruby is terrible, however)

I reinstalled a clean chefdk 13:
Test Kitchen version 1.7.3

I'm setting this in .kitchen.local.yml
provisioner:
  name: chef_zero
  http_proxy: ""
  https_proxy: ""
  chef_client_path: '/usr/bin/chef-client'
  require_chef_omnibus: false

and I'm seeing this in the kitchen converge -l debug for all of the ssh calls. 

D      [SSH] reusing existing connection cloud-user@192.168.4.222<{:user_known_hosts_file=>"/dev/null", :paranoid=>false, :port=>22, :compression=>true, :compression_level=>6, :keepalive=>true, :keepalive_interval=>60, :timeout=>10, :user=>"cloud-user"}>
D      [SSH] cloud-user@192.168.4.222<{:user_known_hosts_file=>"/dev/null", :paranoid=>false, :port=>22, :compression=>true, :compression_level=>6, :keepalive=>true, :keepalive_interval=>60, :timeout=>10, :user=>"cloud-user"}> (sh -c '
http_proxy="http://myinternalproxy.mycompany.com:3128"; export http_proxy
https_proxy="http://myinternalproxy.mycompany.com:3128"; export https_proxy
sudo_rm="sudo -E rm"
dirs="/tmp/kitchen/clients /tmp/kitchen/cookbooks /tmp/kitchen/data /tmp/kitchen/data_bags /tmp/kitchen/encrypted_data_bag_secret /tmp/kitchen/environments /tmp/kitchen/roles"
root_path="/tmp/kitchen"

**Comment by smurawski** [comment]
This should be resolved via https://github.com/test-kitchen/test-kitchen/pull/1027 which was shipped in v1.9.0
